### PR TITLE
Remove remaining instances of proto.clone() 

### LINKED ISSUE #4757: Eliminate All Uses of Proto.Clone From Prysm
Proto.clone is an extremely expensive method that we have taken great lengths to avoid in some recent refactors. However, a quick grep in Prysm shows it still being used significantly for various reasons, especially in certain helper functions and cached values in the beacon chain. Now, we have a bunch of custom copy functions in `beacon-chain/state/getters.go` that we can use instead of an expensive, generic, proto.Clone call.

This is a tracking issue for us to use the custom copy functions from `beacon-chain/state/getters.go` instead of proto.Clone throughout the repo.

### PR DESCRIPTION
**Pull request by garyschulteog**
TODO(you): choose "part of" or "resolves" and the associated github issue #.

Resolves #4757

---

# Description

* use existing Copy* methods from state/getters.go 
* add Copy* methods to state/getters.go to remove proto.Clone() from initial-sync
* resolve test breakage fallout
   * most breakage seems due to the different resulting object between proto.Clone and the Copy* methods; the latter yields deeper copies of sparse objects which serialize differently

**Write why you are making the changes in this pull request**

**Write a summary of the changes you are making**

**Link anything that would be helpful or relevant to the reviewers**

### CODE REVIEW

**Comment by terencechain** [comment]
Thanks for opening this up!

**Comment by prestonvanloon** [comment]
Thanks for doing this @garyschulte. Can I recommend that we merge this PR incrementally? If you have some changes that are ready to go, lets put this PR for review and continue progress in a follow up PR.

**Comment by garyschulteog** [comment]
> Thanks for doing this @garyschulte. Can I recommend that we merge this PR incrementally? If you have some changes that are ready to go, lets put this PR for review and continue progress in a follow up PR.

I removed the use of CopySignedBeaconBlock in `ReceiveBlockNoVerify` for now.  `reflect.DeepEquals` reports that objects cloned via `CopySignedBeaconBlock(block)` are identical to those cloned with `proto.Clone(block).(*ethpb.SignedBeaconBlock)`.  However, during testing, intial sync always panics at slot 860 during testing if using the copy method rather than proto.Clone there.  🤷‍♂ 

I am have been running this branch for a few hours on the test net on a synced node, and am have nearly completely synced from 0 on another machine with this branch.    To the best of my knowledge it is safe, but I honestly don't understand the root cause for the panics when using that clone method during sync.

File: beacon-chain/blockchain/receive_block.go [hunk 1]
```diff
@@ -183,7 +184,7 @@ func (s *Service) ReceiveBlockNoPubsubForkchoice(ctx context.Context, block *eth
 func (s *Service) ReceiveBlockNoVerify(ctx context.Context, block *ethpb.SignedBeaconBlock) error {
 	ctx, span := trace.StartSpan(ctx, "beacon-chain.blockchain.ReceiveBlockNoVerify")
 	defer span.End()
-	blockCopy := proto.Clone(block).(*ethpb.SignedBeaconBlock)
+	blockCopy := proto.Clone(block).(*ethpb.SignedBeaconBlock) //stateTrie.CopySignedBeaconBlock(block)
```
Comment (garyschulteog) [inline beacon-chain/blockchain/receive_block.go]: this is the confounding issue mentioned in comments.  Copy method commented out for now and will revisit on a subsequent PR

Comment (garyschulte) [reply beacon-chain/blockchain/receive_block.go]: Thanks again @nisdas for pointing out the proof copy issue.

File: beacon-chain/state/getters.go [hunk 2]
```diff
@@ -647,39 +627,269 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+		Body:                 CopyBeaconBlockBody(block.Body),
+	}
+}
+
+// CopyBeaconBlockBody copies the provided BeaconBlockBody
+func CopyBeaconBlockBody(body *ethpb.BeaconBlockBody) *ethpb.BeaconBlockBody {
+	if body == nil {
+		return nil
+	}
+
```
Comment (rauljordan) [inline beacon-chain/state/getters.go]: Eliminate white space

File: beacon-chain/state/getters.go [hunk 3]
```diff
@@ -647,39 +627,269 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+
+
+	return &ethpb.BeaconBlockBody{
+		RandaoReveal: 		  safeCopyBytes(body.RandaoReveal),
```
Comment (rauljordan) [inline beacon-chain/state/getters.go]: Needs to run gofmt in the project

File: beacon-chain/state/getters.go [hunk 4]
```diff
@@ -647,39 +627,269 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+		Graffiti:             safeCopyBytes(body.Graffiti),
+		ProposerSlashings:    CopyProposerSlashings(body.ProposerSlashings),
+		AttesterSlashings:    CopyAttesterSlashings(body.AttesterSlashings),
+		Attestations:         CopyAttestations(body.Attestations),
+		Deposits:             CopyDeposits(body.Deposits),
+		VoluntaryExits:       CopySignedVoluntaryExits(body.VoluntaryExits),
+	}
+}
+
+// CopyProposerSlashings copies the provided ProposerSlashing array
```
Comment (rauljordan) [inline beacon-chain/state/getters.go]: Please end all sentences with proper punctuation.

**Comment by 0xKiwi** [comment]
Fixed the suggestions for ya @garyschulte  😄

File: beacon-chain/state/getters.go [hunk 5]
```diff
@@ -647,39 +628,256 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+	return &ethpb.SignedVoluntaryExit{
+		Exit: &ethpb.VoluntaryExit{
+			Epoch:          exit.Exit.Epoch,
+			ValidatorIndex: exit.Exit.ValidatorIndex,
+		},
+		Signature: safeCopyBytes(exit.Signature),
+	}
+}
+
+func safeCopyBytes(cp []byte) []byte {
```
Comment (nisdas) [inline beacon-chain/state/getters.go]: hey could you maybe put this  in 
```
/shared/bytesutil
```
this is a useful helper that can be used across the repo.

File: beacon-chain/state/getters.go [hunk 6]
```diff
@@ -647,39 +628,256 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+	if cp != nil {
+		copied := make([]byte, len(cp))
+		copy(copied, cp)
+		return copied
+	}
+	return nil
+}
+
+func copy2dBytes(ary [][]byte) [][]byte {
```
Comment (nisdas) [inline beacon-chain/state/getters.go]: same for this one, we have a lot of 2d byte arrays that we copy

File: beacon-chain/state/getters.go [hunk 7]
```diff
@@ -647,39 +628,256 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
+	copied := make([][]byte, len(ary))
+
+	for i := 0; i < len(ary); i++ {
```
Comment (nisdas) [inline beacon-chain/state/getters.go]:
```suggestion
		for i,a := range ary {
		copied[i] = safeCopyBytes(a)
	}

```
should be more straightforward

File: beacon-chain/state/getters.go [hunk 8]
```diff
@@ -647,39 +628,256 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
...
-
 	return &ethpb.Checkpoint{
 		Epoch: cp.Epoch,
-		Root:  root[:],
+		Root:  safeCopyBytes(cp.Root),
+	}
+}
+
+// CopySignedBeaconBlock copies the provided SignedBeaconBlock.
+func CopySignedBeaconBlock(sigBlock *ethpb.SignedBeaconBlock) *ethpb.SignedBeaconBlock {
```
Comment (nisdas) [inline beacon-chain/state/getters.go]: @garyschulteog how about putting all copy methods in a separate file ?

File: beacon-chain/state/getters.go (hunk 5 above)
Comment (garyschulteog) [reply beacon-chain/state/getters.go]: Moved :)

**Pull request merged**